### PR TITLE
bluetooth: mesh: revisions to support building C++ applications

### PR DESCRIPTION
**Pull request by pabigot**
A couple changes to the headers to avoid violating C++ language requirements.

Using the mesh access macros in C++ also requires #25593.

### CODE REVIEW

**Comment by pabigot** [comment]
Draft until #25593 merged.

File: include/bluetooth/mesh/access.h
Comment (jhedberg) [inline include/bluetooth/mesh/access.h:261]: This isn't doing the same thing as the original code. `BT_MESH_KEY_UNUSED` is defined as `0xffff`, however what you're doing here is only initialising `.keys[0]` to that value, whereas all other values of the array get initialised to zero.

Comment (pabigot) [reply include/bluetooth/mesh/access.h:261]: Thanks; I've never used that syntax and didn't realize it was a spread operation.  If so, there may be no way to make this header C++ compatible.

Comment (jhedberg) [reply include/bluetooth/mesh/access.h:261]: It's a GNU C extension: https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html

However, it's a fair question to ask whether we should allow such extensions or require standard C code everywhere. It's super convenient though and probably saves a bit of CPU time since the array doesn't need to be initialised at runtime.

Comment (joerchan) [reply include/bluetooth/mesh/access.h:261]: I haven't tested it out, but I think the designated initializer can be replaced by some of the macro utils (include/sys/util.h) that we have.

Something like:
```
#define FOO(i, _) BT_MESH_ADDR_UNASSIGNED ,
.groups = { UTIL_LISTIFY(CONFIG_BT_MESH_MODEL_GROUP_COUNT - 1, FOO) }
```
I'm not expert on these though, so you might find a better way in there.

Comment (pabigot) [reply include/bluetooth/mesh/access.h]: @joerchan Thanks, that seems to work.  It does require defining a helper macro to generate the constant; maintainers please provide feedback on how that should be spelled and hidden from the user.

One approach not done here is to move the helper functions to the bottom of the file, outside the doxygen block.  This works, but may be confusing.

Comment (joerchan) [reply include/bluetooth/mesh/access.h]: I think @trond-snekvik is the best person to answer the naming and hiding question. I suppose this should probably be re-usable by users for their own models and so it shouldn't be hidden.

Comment (pabigot) [reply include/bluetooth/mesh/access.h]: OK.  I provided a new one that puts all the internals above the documentation block.  The one with "public" (no `_` affix) could be moved into the block and documented, if it provides something of use outside the helper macros.

**Comment by carlescufi** [comment]
@pabigot do you want to move this out of draft?

**Review by trond-snekvik — APPROVED** [review]

**Review by tsvehagen — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
